library/L1_Peripheral/msp432p401r/system_controller.hpp

```diff
@@ -376,91 +514,156 @@ class SystemController final : public sjsu::SystemController
+      uint8_t dco_frequency_select = 0b000;
+      float dco_constant =
+          static_cast<float>(device_descriptors->DCOIR_CONSTK_RSEL04);
+      int32_t dco_calibration =
+          static_cast<int32_t>(device_descriptors->DCOIR_FCAL_RSEL04);
+
+      if (kTargetFrequency >= 32_MHz)
+      {
+        // target frequency is 32 MHz to 48 MHz
+        dco_frequency_select = 0b101;
+        dco_constant =
+            static_cast<float>(device_descriptors->DCOIR_CONSTK_RSEL5);
+        dco_calibration =
+            static_cast<int32_t>(device_descriptors->DCOIR_FCAL_RSEL5);
+      }
+      else if (kTargetFrequency >= 16_MHz)
+      {
+        // target frequency is 16 MHz to 32 MHz
+        dco_frequency_select = 0b100;
+      }
+      else if (kTargetFrequency >= 8_MHz)
+      {
+        // target frequency is 8 MHz to 16 MHz
+        dco_frequency_select = 0b011;
+      }
+      else if (kTargetFrequency >= 4_MHz)
+      {
+        // target frequency is 4 MHz to 8 MHz
+        dco_frequency_select = 0b010;
+      }
+      else if (kTargetFrequency >= 2_MHz)
+      {
+        // target frequency is 2 MHz to 4 MHz
+        dco_frequency_select = 0b001;
+      }
+      else
+      {
+        // target frequency is 1 MHz to 2 MHz with center frequency of 1.5 MHz
+        // use the default initialized values
+      }
+
+      // =======================================================================
+      // Step 3. Calculate the signed 10-bit tuning value using Equation 6 from
+      //         https://www.ti.com/lit/ug/slau356i/slau356i.pdf#page=387
+      // =======================================================================
+      constexpr units::frequency::kilohertz_t kDcoCenterFrequencies[] = {
+        1500_kHz, 3_MHz, 6_MHz, 12_MHz, 24_MHz, 48_MHz
+      };
+      const auto kCenterFrequency = kDcoCenterFrequencies[dco_frequency_select];
+      const float kFrequencyDifference =
+          (kTargetFrequency - kCenterFrequency).to<float>();
+      const float kCalibration =
+          (1.0f + dco_constant * static_cast<float>(768 - dco_calibration));
+      const int16_t kTuneValue =
+          static_cast<int16_t>((kFrequencyDifference * kCalibration) /
+                               (kTargetFrequency.to<float>() * dco_constant));
+
+      // =======================================================================
+      // Step 4. Configure the CSCTL0 register.
+      // =======================================================================
+      UnlockClockSystemRegisters();
+      {
+        clock_system->CTL0 = bit::Insert(clock_system->CTL0, kTuneValue,
+                                         Control0Register::kTuningSelect);
+        clock_system->CTL0 =
+            bit::Insert(clock_system->CTL0, dco_frequency_select,
+                        Control0Register::kFrequencySelect);
+        clock_system->CTL0 =
+            bit::Set(clock_system->CTL0, Control0Register::kEnable);
```

Comment:
It would be better to read CTL0 into a variable, and perform the insertion operations on the variable to allow for easier optimization by the compiler. Unless these steps must be done in this specific order, in which case, add a comment to contextualize that.